### ISSUE TITLE
Support ARM64EC on Windows

### ISSUE THREAD

**Issue by thhous-msft** [openssl/openssl#16482]
Arm64EC is a new arm64 calling convention in Windows 11 to make mixing x64 and arm64 code easier. Theoretically, this should just be as easy as adding a new configuration copying the existing arm64 windows configuration. I'm working on testing this now, and if it works I'll PR it.

This should be looked at as a candidate to backport to 1.1.1. Similar to the arm64 osx discussion, it requires no code changes, just a configuration change.

**Comment by davidben** [comment]
OpenSSL has a lot of assembly code, which is written to the standard AAPCS64 calling convention. How does this new calling convention interact those? Do assembly function prototypes need to be marked so that ARM64EC -> AAPCS64 transitions work correctly? Is there any documentation on what exactly the ARM64EC calling convention is (which arguments in which registers, etc.)?

**Comment by thhous-msft** [comment]
Based on what I was looking at, it doesn't actually look like the windows targets in Openssl currently use any of the arm64 assembly code. 

As for the ABI itself, I'll try to see if I can find some external documentation. But generally, its designed to be a 1 to 1 mapping matching the x64 ABI.

**Comment by davidben** [comment]
> Based on what I was looking at, it doesn't actually look like the windows targets in Openssl currently use any of the arm64 assembly code.

I don't believe that's true in OpenSSL 3.0. Though, sure, one could always say ARM64EC does not get the benefit of assembly optimizations, due to its non-standard calling convention. Short work like something below, that's probably the most practical answer.

> But generally, its designed to be a 1 to 1 mapping matching the x64 ABI.

Right, I've seen that before, but that doesn't describe a calling convention. It's just a vague description of the design goals. Documentation should tell you exactly which registers are used where, etc., otherwise no one can write code to that calling convention.

Although, to avoid imposing unreasonable costs on the rest of the ecosystem (supporting two calling conventions in one assembly file is a significant maintenance burden), I would encourage you all to make ARM64EC -> AAPCS64 calls work. Assembly functions are typically internal to a library, rather than exposed directly. That means, you don't need to worry about calls from emulated-x64, only ARM64EC. With suitable attributes in your compiler, you could have assembly functions stick to the standard calling convention, and then the non-standard callers can bridge that accordingly.

**Comment by thhous-msft** [comment]
I'm not part of the compiler team at all. I just work on our QUIC library, and am trying to get it to build for arm64ec. I can attempt to pass that information off to the right compiler teams however.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by RashmiRamanna4** [comment]
Is this issue resolved? Do we have support for ARM64EC platform ?

**Comment by joshyoo-topazlabs** [comment]
I would love some support on this as well

**Comment by quic-zhanweiw** [comment]
I can compile openssl v3.4.0 to ARM64EC successfully with the configuration 'no-asm'. But I think it will impact the openssl performance obviously. Hope can support ARM64EC officially with assembly code enabled.
Here has the detailed information about calling convention of ARM64EC:
https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by tregubovav-dev** [comment]
I have successfully compiled `OpenSSL` for the Windows `arm64ec` target. The `openssl.exe` executable runs basic commands correctly, `arm64ec` applications can load the shared library and use its routines, and the full `OpenSSL` test suite passes.

Based on this, I suggest declaring `experimental` support for this target. However, there is a critical ambiguity regarding library naming that needs to be addressed first.

Windows on ARM64 currently supports three distinct execution models, making the naming convention vital for side-by-side installations:

1.  **x64 Emulation**: Runs standard `x64` binaries (`openssl.exe`, `libcrypto-3-x64.dll`) via the OS emulation layer. This is compatible with all Win64 x64 apps but incurs a significant performance penalty.
2.  **ARM64EC (Emulation Compatible)**: Runs native ARM64 code but maintains interoperability with the x64 ABI. These binaries can load and be loaded by x64 DLLs/processes.
3.  **Native ARM64**: Runs pure ARM64 code (AArch64). These binaries are incompatible with x64/ARM64EC processes.
4.  **ARM64X**: A hybrid binary format containing both ARM64 and ARM64EC code, allowing the OS to load the appropriate section automatically.

**The Issue:**
Currently, OpenSSL defaults to the `-arm64` suffix for ARM-based Windows builds. If we introduce a pure `arm64ec` build target without a distinct naming scheme, it may produce libraries (e.g., `libcrypto-3-arm64.dll`) that conflict with native ARM64 libraries. This prevents applications from distinguishing between the Native ARM64 ABI and the ARM64EC ABI at the file level.

**Suggestion:**
I propose adding a dedicated (experimental) build target for `arm64ec` that automatically appends a unique suffix to the shared libraries, such as `-arm64ec` (e.g., `libcrypto-3-arm64ec.dll`). This follows the existing precedent of the `-x64` and `-arm64` suffixes and ensures that `arm64ec` artifacts can coexist with native builds.

**Comment by Myriachan** [comment]
Note that the ARM64EC ABI prohibits the use of the `x13`, `x14`, `x23`, `x24`, `x28`, and `v16`-`v31` registers.  It might be necessary to disable ARM64 assembly language files in ARM64EC builds.

https://learn.microsoft.com/en-us/cpp/build/arm64ec-windows-abi-conventions?view=msvc-170

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
